splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/window/function/StringAggregator.java

```diff
@@ -0,0 +1,108 @@
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.derby.impl.sql.execute.operations.window.function;
+
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.services.io.FormatableHashtable;
+import com.splicemachine.db.iapi.services.loader.ClassFactory;
+import com.splicemachine.db.iapi.sql.execute.WindowFunction;
+import com.splicemachine.db.iapi.types.DataTypeDescriptor;
+import com.splicemachine.db.iapi.types.DataValueDescriptor;
+import com.splicemachine.db.iapi.types.NumberDataValue;
+import com.splicemachine.db.iapi.types.SQLLongint;
+import com.splicemachine.db.iapi.types.SQLVarchar;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+
+/**
+ *
+ * Created by jyuan on 7/31/14.
+ */
+public class StringAggregator extends SpliceGenericWindowFunction {
+    private StringBuilder acc;
+    private String separator;
+
+    public StringAggregator(){}
+
+    public StringAggregator(String separator) {
+        this.separator = separator;
+    }
+
+    @Override
+    public WindowFunction setup(ClassFactory classFactory, String aggregateName, DataTypeDescriptor returnDataType,
+                                FormatableHashtable functionSpecificArgs) throws StandardException {
+        super.setup( classFactory, aggregateName, returnDataType );
+        acc = new StringBuilder();
+        separator = ((DataValueDescriptor) functionSpecificArgs.get("param")).getString();
+        return this;
+    }
+
+    @Override
+    public void accumulate(DataValueDescriptor[] valueDescriptors) throws StandardException {
+        this.add(valueDescriptors);
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        acc = new StringBuilder();
+    }
+
+    @Override
+    protected void calculateOnAdd(WindowChunk chunk, DataValueDescriptor[] dvds) throws StandardException {
+        if (acc.length() == 0)
+            acc.append(dvds[0].getString());
+        else
+            acc.append(separator).append(dvds[0].getString());
+    }
+
+    @Override
+    protected void calculateOnRemove(WindowChunk chunk, DataValueDescriptor[] dvds) throws StandardException {
+        String str = dvds[0].getString();
+        int from = acc.indexOf(str);
+        if (from == -1) return;
+        int len = str.length() + separator.length();
+        if (len > acc.length()) len = acc.length();
+        acc.delete(from, len);
+    }
+
+    @Override
+    public DataValueDescriptor getResult() throws StandardException {
+        return new SQLVarchar(acc.toString());
+    }
+
+    @Override
+    public WindowFunction newWindowFunction() {
+        return new StringAggregator(separator);
+    }
+
+    @Override
+    public void writeExternal(ObjectOutput out) throws IOException {
+        out.writeObject(acc);
```

Comment:
Please consider writing a class version number (e.g. 1) as part of SerDe in case this is every stored on disk.